### PR TITLE
Sample of using an external directory for pages

### PR DESCRIPTION
**Pull request by rynowak**
Sorry for the confusion, I think I misunderstood what you were asking for.

### CODE REVIEW

File: src/DummyFileBasedCms.Web/MvcInfrastructureStuff/FlatFileCmsRazorProjectPageRouteModelProvider.cs [hunk 1]
```diff
@@ -1,145 +0,0 @@
...
-    using Microsoft.AspNetCore.Hosting;
-    using Microsoft.AspNetCore.Mvc.ApplicationModels;
-    using Microsoft.AspNetCore.Mvc.RazorPages;
-    using Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure;
-    using Microsoft.AspNetCore.Mvc.RazorPages.Internal;
-    using Microsoft.AspNetCore.Razor.Language;
-    using Microsoft.Extensions.Logging;
-    using Microsoft.Extensions.Options;
-
-    public class FlatFileCmsRazorProjectPageRouteModelProvider : IPageRouteModelProvider
```
Comment (rynowak) [inline src/DummyFileBasedCms.Web/MvcInfrastructureStuff/FlatFileCmsRazorProjectPageRouteModelProvider.cs:15]: Implementing an `IPageRouteModelProvider` might still be useful if you want to customize how pages in the external directory behave. In `OnProvidersExecuted` you should be able to see and modify what was created by the default providers.

File: src/DummyFileBasedCms.Web/Startup.cs [hunk 2]
```diff
@@ -25,24 +19,37 @@ public Startup( IConfiguration configuration )
...
 
-            services.AddSingleton(fileProvider);
-            services.AddTransient<FlatFileCmsProviderRazorProjectFileSystem>();
-            services.AddSingleton<IPageRouteModelProvider, FlatFileCmsRazorProjectPageRouteModelProvider>();
+                // File Providers throw if the directory doesn't exist. 
+                if (Directory.Exists(cmsOptions.FilePath))
+                {
+                    options.FileProviders.Add(new PhysicalFileProvider(cmsOptions.FilePath));
+                }
+            });
```
Comment (rynowak) [inline src/DummyFileBasedCms.Web/Startup.cs:45]: This is all that's required to get MVC to look in another directory - if you need to change directories at runtime, thats a case where you might need to implement your own file provider.

**Comment by rynowak** [comment]
Let me know if this does what you want 👍

**Comment by sparkeh9** [comment]
@rynowak This works as I'd hoped - I didn't have any expectations of anyone getting so involved so your help is very much appreciated, thank you.

A few observations from trying to get this to work just now:
- When I set the folder structure in the configuration, it didn't work unless I had the pages in a "Pages" folder, without including that in the config
- When I included the additional physical file provider, the ones included under the app root stopped resolving.  I tried adding a second physical file provider (so we're up to three now including the app root one) and only the first one added seems to allow pages to be resolved
- Inside the `services.Configure<RazorViewEngineOptions>` block, options.FileProviders is an empty collection, it seems the built-in provider is added some time after this stage, because when I inject `IRazorViewEngineFileProviderAccessor`, I see all the providers which I added, with the built-in one at the last index.  Only the first in the collection is looked at and serves pages

**Comment by sparkeh9** [comment]
image

**Comment by rynowak** [comment]
OK thanks, I'll take a look at this and see what's missing.

**Comment by rynowak** [comment]
Sorry for the delay on this, I'll try to get to it tonite.

**Pull request merged**